src/session_impl.cpp

```diff
@@ -2448,6 +2465,53 @@ namespace {
 		m_i2p_listen_socket.reset();
 		open_new_incoming_i2p_connection();
 	}
+
+	void session_impl::schedule_i2p_reconnect()
+	{
+		if (m_abort) return;
+		
+		// If I2P is not configured, don't schedule reconnection
+		if (m_settings.get_str(settings_pack::i2p_hostname).empty()) return;
+		
+		// Already reconnecting
+		if (m_i2p_reconnecting) return;
+		
+#ifndef TORRENT_DISABLE_LOGGING
+		if (should_log())
+			session_log("scheduling i2p reconnection in %d seconds", m_i2p_reconnect_delay);
+#endif
+		
+		m_i2p_reconnecting = true;
+		
+		// Cancel any existing timer
+		m_i2p_reconnect_timer.cancel();
+		
+		// Schedule reconnection with current delay
+		m_i2p_reconnect_timer.expires_after(seconds(m_i2p_reconnect_delay));
+		m_i2p_reconnect_timer.async_wait(
+			std::bind(&session_impl::on_i2p_reconnect_timer, this, _1));
```

Comment:
I think it might be safer to pass in a weak reference to the session impl here, rather than `this`. Something like this:

```
std::weak_ptr<session_impl> self = shared_from_this();
m_i2p_reconnect_timer.async_wait(
	std::bind(&on_i2p_reconnect_timer, std::move(self), _1));
```

The handler must then take a `weak_ptr`, promote it to a `shared_ptr`, and only if that's successful, actually call the member function `on_i2p_reconnect_timer()`